File: app/src/main/java/com/example/maksim/testapp/activities/MainActivity.java
package com.example.maksim.testapp.activities;

import android.app.Fragment;
import android.app.FragmentManager;
import android.app.FragmentTransaction;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.os.Parcelable;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;

import com.example.maksim.testapp.R;
import com.example.maksim.testapp.fragments.DetailsFragment;
import com.example.maksim.testapp.fragments.ListFragment;

public class MainActivity extends AppCompatActivity
        implements com.example.maksim.testapp.fragments.ListFragment.OnListFragmentInteractionListener {

    private final String LOG = "MainActivity";
    private FragmentManager fragmentManager;
    private ActionBar actionBar;
    private boolean isLandTablet;
    private String savedGitHubUserState;
    private Parcelable savedRecyclerLayoutState;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        Log.e(LOG, "onCreate");
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        init(savedInstanceState);
    }

    private void init(Bundle savedInstanceState) {
        isLandTablet = getResources().getBoolean(R.bool.isLandTablet);
        fragmentManager = getFragmentManager();

        actionBar = getSupportActionBar();
        setFragmentManagerListener();

        if(savedInstanceState != null) {
            savedRecyclerLayoutState = savedInstanceState
                    .getParcelable(ListFragment.RECYCLER_LAYOUT_STATE);
            savedGitHubUserState = savedInstanceState
                    .getString(DetailsFragment.SELECTED_MODEL);
        }

        if (savedInstanceState == null && !isLandTablet) {
            ListFragment listFragment = new ListFragment();
            startFragment(listFragment, R.id.fragmentContainer, true, false);
        } else if(savedInstanceState != null) {
            if (!isLandTablet) {
                updatePortraitTabletAndPhoneState();
            } else {
                updateLandTabletState();
            }
        }

        setActionBarTitle();
    }

    private void updateLandTabletState() {
        clearBackStack();

        ListFragment listFragment = (ListFragment) fragmentManager
                .findFragmentById(R.id.list_fragment);
        if (listFragment != null) {
            listFragment.setSavedRecyclerLayoutState(savedRecyclerLayoutState);
        }

        DetailsFragment detailsFragment = (DetailsFragment) fragmentManager
                .findFragmentById(R.id.details_fragment);
        if (detailsFragment != null) {
            detailsFragment.updateContent(savedGitHubUserState);
        }
    }

    private void updatePortraitTabletAndPhoneState() {
        Fragment fragment = fragmentManager.findFragmentById(R.id.fragmentContainer);
        if (fragment != null) {
            if(fragment instanceof ListFragment) {
                ((ListFragment) fragment).setSavedRecyclerLayoutState(savedRecyclerLayoutState);
            } else if(fragment instanceof DetailsFragment) {
                ((DetailsFragment) fragment).updateContent(savedGitHubUserState);
            }
        } else {
            Bundle bundle = new Bundle();
            bundle.putParcelable(ListFragment.RECYCLER_LAYOUT_STATE, savedRecyclerLayoutState);

            ListFragment listFragment = new ListFragment();
            listFragment.setArguments(bundle);
            startFragment(listFragment, R.id.fragmentContainer, true, false);
        }
    }

    @Override
    protected void onSaveInstanceState(final Bundle outState) {
        super.onSaveInstanceState(outState);

        if(isLandTablet) {
            saveStateFromLandTablet(outState);
        } else {
            saveStateFromPortraitTabletAndPhone(outState);
        }
    }

    private void saveStateFromLandTablet(Bundle outState) {
        ListFragment listFragment = (ListFragment) fragmentManager.findFragmentById(R.id.list_fragment);
        if (listFragment != null) {
            savedRecyclerLayoutState = listFragment.getSavedRecyclerLayoutState();
        }
        outState.putParcelable(ListFragment.RECYCLER_LAYOUT_STATE, savedRecyclerLayoutState);

        DetailsFragment detailsFragment = (DetailsFragment) fragmentManager
                .findFragmentById(R.id.details_fragment);
        if (detailsFragment != null) {
            String savedGitHubUserState = detailsFragment.getGitHubUser();
            outState.putString(DetailsFragment.SELECTED_MODEL, savedGitHubUserState);
        }
    }

    private void saveStateFromPortraitTabletAndPhone(Bundle outState) {
        Fragment fragment = fragmentManager.findFragmentById(R.id.fragmentContainer);
        if (fragment != null) {
            if(fragment instanceof ListFragment) {
                savedRecyclerLayoutState = ((ListFragment) fragment).getSavedRecyclerLayoutState();
            } else if(fragment instanceof DetailsFragment) {
                savedGitHubUserState = ((DetailsFragment) fragment).getGitHubUser();
            }

            outState.putParcelable(ListFragment.RECYCLER_LAYOUT_STATE, savedRecyclerLayoutState);
            outState.putString(DetailsFragment.SELECTED_MODEL, savedGitHubUserState);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_main, menu);

        if(!isLandTablet) {
            setMenuBackButton();
        }
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                if (fragmentManager.getBackStackEntryCount() > 0) {
                    fragmentManager.popBackStack();
                }
                break;
            case R.id.menu_edit_query:
                showEditQueryDialog();
                break;
            default:
        }
        return super.onOptionsItemSelected(item);
    }

    private void showEditQueryDialog() {
        AlertDialog.Builder dialogBuilder = new AlertDialog.Builder(this);
        LayoutInflater inflater = this.getLayoutInflater();
        final View dialogView = inflater.inflate(R.layout.edit_dialog, null);
        dialogBuilder.setView(dialogView);

        final EditText editText = (EditText) dialogView.findViewById(R.id.editText);
        SharedPreferences prefs =
                getSharedPreferences(ListFragment.PREFERENCE_SEARCH_QUERY, MODE_PRIVATE);
        String searchQuery = prefs.getString(ListFragment.SEARCH_QUERY, "");
        editText.setText(searchQuery);

        dialogBuilder.setTitle(getString(R.string.edit_dialog_title));
        dialogBuilder.setPositiveButton("Done", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                String searchQuery = editText.getText().toString();
                saveQueryText(searchQuery);
            }
        });
        dialogBuilder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {

            }
        });
        AlertDialog b = dialogBuilder.create();
        b.show();
    }

    private void saveQueryText(String searchQuery) {
        SharedPreferences.Editor editor =
                getSharedPreferences(ListFragment.PREFERENCE_SEARCH_QUERY, MODE_PRIVATE).edit();
        editor.putString(ListFragment.SEARCH_QUERY, searchQuery);
        editor.apply();

        ListFragment listFragment = (ListFragment) fragmentManager
                .findFragmentById(R.id.list_fragment);
        if (listFragment == null) {
            listFragment = (ListFragment) fragmentManager
                    .findFragmentById(R.id.fragmentContainer);
        }

        if (listFragment != null)
            listFragment.executeRequest();
    }

    private void setFragmentManagerListener() {
        fragmentManager.addOnBackStackChangedListener(new FragmentManager.OnBackStackChangedListener() {
            @Override
            public void onBackStackChanged() {
                setMenuBackButton();
            }
        });
    }

    private void setMenuBackButton() {
        if(actionBar != null) {
            int backStackCount = fragmentManager.getBackStackEntryCount();
            boolean enableBackButton = backStackCount > 0;
            actionBar.setHomeButtonEnabled(enableBackButton);
            actionBar.setDisplayHomeAsUpEnabled(enableBackButton);
        }

        setActionBarTitle();
    }

    private void setActionBarTitle() {
        if(isLandTablet) {
            setActionBarTitle(getString(R.string.menu_title_list_and_details));
        } else {
            Fragment fragment = fragmentManager.findFragmentById(R.id.fragmentContainer);
            if (fragment instanceof DetailsFragment) {
                setActionBarTitle(getString(R.string.menu_title_details));
            } else {
                setActionBarTitle(getString(R.string.menu_title_list));
            }
        }
    }

    private void setActionBarTitle(String title) {
        Log.e(LOG, "setActionBarTitle " + title);
        if(actionBar != null)
            actionBar.setTitle(title);
    }

    private void startFragment(Fragment fragment, int fragmentContainer, boolean replace, boolean addToBackStack) {
        FragmentTransaction transaction = fragmentManager.beginTransaction();
        if (replace) {
            transaction = transaction.replace(fragmentContainer, fragment);
        } else {
            transaction = transaction.add(fragmentContainer, fragment);
        }
        if(addToBackStack)
            transaction.addToBackStack(null);
        transaction.commitAllowingStateLoss();
    }

    private void clearBackStack() {
        if (fragmentManager.getBackStackEntryCount() > 0) {
            FragmentManager.BackStackEntry backStackEntry = fragmentManager.getBackStackEntryAt(0);
            fragmentManager.popBackStack(backStackEntry.getId(), FragmentManager.POP_BACK_STACK_INCLUSIVE);
        }
    }

    @Override
    public void onListFragmentInteraction(String userLogin) {
        Log.e(LOG, "onListFragmentInteraction");
        savedGitHubUserState = userLogin;
        onItemSelected(userLogin);
    }

    @Override
    public void setFirstElementOfList(String userLogin) {
        DetailsFragment detailsFragment = (DetailsFragment) fragmentManager
                .findFragmentById(R.id.details_fragment);
        if (detailsFragment != null && isLandTablet)
            detailsFragment.updateContent(userLogin);
    }

    private void onItemSelected(String userLogin) {
        DetailsFragment detailsFragment = (DetailsFragment) fragmentManager
                .findFragmentById(R.id.details_fragment);
        Bundle bundle = new Bundle();
        bundle.putString(DetailsFragment.SELECTED_MODEL, userLogin);
        if (detailsFragment == null || !isLandTablet) {
            ListFragment listFragment = (ListFragment) fragmentManager.findFragmentById(R.id.fragmentContainer);
            if(listFragment != null)
                savedRecyclerLayoutState = listFragment.getSavedRecyclerLayoutState();

            detailsFragment = new DetailsFragment();
            detailsFragment.setArguments(bundle);
            startFragment(detailsFragment, R.id.fragmentContainer, false, true);

            setActionBarTitle(getString(R.string.menu_title_details));
        } else {
            detailsFragment.updateContent(userLogin);
        }
    }
}

File: app/src/main/java/com/example/maksim/testapp/activities/MainActivity.kt
package com.example.maksim.testapp.activities

import android.app.Fragment
import android.app.FragmentManager
import android.content.Context
import android.os.Parcelable
import android.support.v7.app.ActionBar
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.EditText

import com.example.maksim.testapp.R
import com.example.maksim.testapp.details.fragment.DetailsFragment
import com.example.maksim.testapp.list.fragment.ListFragment

class MainActivity : AppCompatActivity(), ListFragment.OnListFragmentInteractionListener {

    private val LOG = "MainActivity"
    private val fragmentManagerVal: FragmentManager = fragmentManager

    private var actionBar: ActionBar? = null
    private var isLandTablet: Boolean = false
    private var savedGitHubUserData: String? = null
    private var savedRecyclerLayoutState: Parcelable? = null
    private var menuSearchUsers: MenuItem? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        Log.e(LOG, "onCreate")
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        init(savedInstanceState)
    }

    private fun init(savedInstanceState: Bundle?) {
        isLandTablet = resources.getBoolean(R.bool.isLandTablet)
        actionBar = supportActionBar
        setFragmentManagerListener()

        if (savedInstanceState == null && !isLandTablet) {
            val listFragment = ListFragment()
            startFragment(listFragment, R.id.fragmentContainer, true, false)
        } else if (savedInstanceState != null) {
            savedRecyclerLayoutState = savedInstanceState
                    .getParcelable(ListFragment.RECYCLER_LAYOUT_STATE)
            savedGitHubUserData = savedInstanceState
                    .getString(DetailsFragment.SELECTED_MODEL)

            if (!isLandTablet) {
                updatePortraitTabletAndPhoneState()
            } else {
                updateLandTabletState()
            }
        }

        setActionBarTitle()
    }

    private fun setFragmentManagerListener() {
        fragmentManagerVal.addOnBackStackChangedListener { setMenuBackButton() }
    }

    private fun updateLandTabletState() {
        clearBackStack()

        val listFragment = fragmentManagerVal
                .findFragmentById(R.id.listFragment) as ListFragment
        listFragment.setSavedRecyclerLayoutState(savedRecyclerLayoutState)

        val detailsFragment = fragmentManagerVal
                .findFragmentById(R.id.detailsFragment) as DetailsFragment
        detailsFragment.updateContent(savedGitHubUserData)
    }

    private fun updatePortraitTabletAndPhoneState() {
        val fragment = fragmentManagerVal.findFragmentById(R.id.fragmentContainer)
        if (fragment != null) {
            (fragment as? ListFragment)?.setSavedRecyclerLayoutState(savedRecyclerLayoutState) ?: (fragment as? DetailsFragment)?.updateContent(savedGitHubUserData)
        } else {
            val bundle = Bundle()
            bundle.putParcelable(ListFragment.RECYCLER_LAYOUT_STATE, savedRecyclerLayoutState)

            val listFragment = ListFragment()
            listFragment.arguments = bundle
            startFragment(listFragment, R.id.fragmentContainer, true, false)
        }
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)

        if (isLandTablet) {
            saveStateFromLandTablet(outState)
        } else {
            saveStateFromPortraitTabletAndPhone(outState)
        }
    }

    private fun saveStateFromLandTablet(outState: Bundle?) {
        val listFragment = fragmentManagerVal.findFragmentById(R.id.listFragment) as ListFragment?
        if (listFragment != null) {
            savedRecyclerLayoutState = listFragment.getSavedRecyclerLayoutState()
        }
        outState!!.putParcelable(ListFragment.RECYCLER_LAYOUT_STATE, savedRecyclerLayoutState)

        val detailsFragment = fragmentManagerVal
                .findFragmentById(R.id.detailsFragment) as DetailsFragment?
        if (detailsFragment != null) {
            val savedGitHubUserState = detailsFragment.gitHubUser
            outState.putString(DetailsFragment.SELECTED_MODEL, savedGitHubUserState)
        }
    }

    private fun saveStateFromPortraitTabletAndPhone(outState: Bundle?) {
        val fragment = fragmentManagerVal.findFragmentById(R.id.fragmentContainer)
        if (fragment != null) {
            if (fragment is ListFragment) {
                savedRecyclerLayoutState = fragment.getSavedRecyclerLayoutState()
            } else if (fragment is DetailsFragment) {
                savedGitHubUserData = fragment.gitHubUser
            }

            outState!!.putParcelable(ListFragment.RECYCLER_LAYOUT_STATE, savedRecyclerLayoutState)
            outState.putString(DetailsFragment.SELECTED_MODEL, savedGitHubUserData)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_main, menu)

        menuSearchUsers = menu.findItem(R.id.menu_search_users)

        if (!isLandTablet) {
            setMenuBackButton()
        }
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> if (fragmentManagerVal.backStackEntryCount > 0) {
                fragmentManagerVal.popBackStack()
            }
            R.id.menu_search_users -> showEditQueryDialog()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun showEditQueryDialog() {
        val dialogBuilder = AlertDialog.Builder(this)
        val inflater = this.layoutInflater
        val dialogView = inflater.inflate(R.layout.edit_dialog, null)
        dialogBuilder.setView(dialogView)

        val editText = dialogView.findViewById<View>(R.id.editText) as EditText
        val prefs = getSharedPreferences(ListFragment.PREFERENCE_SEARCH_QUERY, Context.MODE_PRIVATE)
        val searchQuery = prefs.getString(ListFragment.SEARCH_QUERY, "")
        editText.append(searchQuery)

        dialogBuilder.setTitle(getString(R.string.edit_dialog_title))
        dialogBuilder.setPositiveButton("Done") { _, _ ->
            saveQueryText(editText.text.toString())
        }
        dialogBuilder.setNegativeButton("Cancel") { _, _ -> }
        val b = dialogBuilder.create()
        b.show()
    }

    private fun saveQueryText(searchQuery: String) {
        val editor = getSharedPreferences(ListFragment.PREFERENCE_SEARCH_QUERY, Context.MODE_PRIVATE).edit()
        editor.putString(ListFragment.SEARCH_QUERY, searchQuery)
        editor.apply()

        if (fragmentManagerVal.backStackEntryCount > 0) {
            fragmentManagerVal.popBackStack()
        }

        var listFragment: ListFragment? = null
        if (!isLandTablet) {
            val fragment = fragmentManagerVal.findFragmentById(R.id.fragmentContainer)
            if (fragment != null && fragment is ListFragment)
                listFragment = fragment
        } else {
            listFragment = fragmentManagerVal
                    .findFragmentById(R.id.listFragment) as ListFragment
        }

        if (listFragment != null)
            listFragment.onSearchQueryChanged()
    }

    private fun setMenuBackButton() {
        if (actionBar != null) {
            val backStackCount = fragmentManagerVal.backStackEntryCount
            val enableBackButton = backStackCount > 0
            actionBar!!.setHomeButtonEnabled(enableBackButton)
            actionBar!!.setDisplayHomeAsUpEnabled(enableBackButton)
        }

        setActionBarTitle()
    }

    private fun setActionBarTitle() {
        if (isLandTablet) {
            setActionBarTitle(getString(R.string.menu_title_list_and_details))
        } else {
            val fragment = fragmentManagerVal.findFragmentById(R.id.fragmentContainer)
            if (fragment is DetailsFragment) {
                setActionBarTitle(getString(R.string.menu_title_details))
            } else {
                setActionBarTitle(getString(R.string.menu_title_list))
            }
        }
    }

    private fun setActionBarTitle(title: String) {
        Log.e(LOG, "setActionBarTitle " + title)
        if (actionBar != null)
            actionBar!!.title = title
    }

    private fun startFragment(fragment: Fragment, fragmentContainer: Int, replace: Boolean, addToBackStack: Boolean) {
        var transaction = fragmentManagerVal.beginTransaction()
        transaction = if (replace)
            transaction.replace(fragmentContainer, fragment)
        else
            transaction.add(fragmentContainer, fragment)
        if (addToBackStack)
            transaction.addToBackStack(null)
        transaction.commitAllowingStateLoss()
    }

    private fun clearBackStack() {
        if (fragmentManagerVal.backStackEntryCount > 0) {
            val backStackEntry = fragmentManagerVal.getBackStackEntryAt(0)
            fragmentManagerVal.popBackStack(backStackEntry.id, FragmentManager.POP_BACK_STACK_INCLUSIVE)
        }
    }

    override fun onListFragmentInteraction(userLogin: String) {
        Log.e(LOG, "onListFragmentInteraction")
        savedGitHubUserData = userLogin
        onItemSelected(userLogin)
    }

    override fun setFirstElementOfList(userLogin: String?) {
        val detailsFragment = fragmentManagerVal
                .findFragmentById(R.id.detailsFragment) as DetailsFragment?
        if (detailsFragment != null && isLandTablet)
            detailsFragment.updateContentIfViewIsEmpty(userLogin!!)
    }

    private fun onItemSelected(userLogin: String) {
        var detailsFragment: DetailsFragment? = fragmentManagerVal
                .findFragmentById(R.id.detailsFragment) as DetailsFragment
        val bundle = Bundle()
        bundle.putString(DetailsFragment.SELECTED_MODEL, userLogin)
        if (detailsFragment == null || !isLandTablet) {
            val listFragment = fragmentManagerVal.findFragmentById(R.id.listFragment) as ListFragment?
            if (listFragment != null)
                savedRecyclerLayoutState = listFragment.getSavedRecyclerLayoutState()

            detailsFragment = DetailsFragment()
            detailsFragment.arguments = bundle
            startFragment(detailsFragment, R.id.fragmentContainer, false, true)

            setActionBarTitle(getString(R.string.menu_title_details))
        } else {
            detailsFragment.updateContent(userLogin)
        }
    }
}
